File: examples/go/p2blake3pkh/P2Blake3PKH_test.go
package contract

import (
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

func TestP2Blake3PKH_Unlock(t *testing.T) {
	pk := runar.Alice.PubKey
	c := &P2Blake3PKH{PubKeyHash: runar.Blake3Hash(pk)}
	c.Unlock(runar.SignTestMessage(runar.Alice.PrivKey), pk)
}

func TestP2Blake3PKH_Unlock_WrongHash(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure for wrong public key hash")
		}
	}()
	pk := runar.Alice.PubKey
	// blake3Hash is mocked (always returns 32 zero bytes), so use a non-matching hash
	wrongHash := runar.ByteString(string(make([]byte, 31)) + "\xff")
	c := &P2Blake3PKH{PubKeyHash: wrongHash}
	c.Unlock(runar.SignTestMessage(runar.Alice.PrivKey), pk)
}

func TestP2Blake3PKH_Compile(t *testing.T) {
	if err := runar.CompileCheck("P2Blake3PKH.runar.go"); err != nil {
		t.Fatalf("Runar compile check failed: %v", err)
	}
}

File: examples/rust/p2blake3pkh/P2Blake3PKH_test.rs
#[path = "P2Blake3PKH.runar.rs"]
mod contract;

use contract::*;
use runar::prelude::*;

#[test]
fn test_unlock() {
    let pk = ALICE.pub_key.to_vec();
    let c = P2Blake3PKH { pub_key_hash: blake3_hash(&pk) };
    c.unlock(&ALICE.sign_test_message(), &pk);
}

#[test]
#[should_panic]
fn test_unlock_wrong_hash() {
    let pk = ALICE.pub_key.to_vec();
    // blake3_hash is mocked (always returns 32 zero bytes), so use a non-matching hash
    let wrong_hash = vec![0xff; 32];
    let c = P2Blake3PKH { pub_key_hash: wrong_hash };
    c.unlock(&ALICE.sign_test_message(), &pk);
}

#[test]
fn test_compile() {
    runar::compile_check(include_str!("P2Blake3PKH.runar.rs"), "P2Blake3PKH.runar.rs").unwrap();
}
